src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,348 @@
+            ( xi_time_event_t* )vector->array[elem_to_cmp_index].selector_t.ptr_value;
+        const xi_time_event_t* elem_to_bubble =
+            ( xi_time_event_t* )vector->array[elem_index].selector_t.ptr_value;
+
+        if ( elem_to_cmp->time_of_execution > elem_to_bubble->time_of_execution )
+        {
+            /* if the elements are not in the right order lets swap them */
+            xi_swap_time_events( vector, elem_to_cmp_index, elem_index );
+
+            /* update the indexes */
+            elem_index -= 1;
+            elem_to_cmp_index -= 1;
+        }
+        else
+        {
+            /* thanks to the container invariant on the order of the elemements we can
+             * break at this moment */
+            break;
+        }
+    }
+
+    return elem_index;
+}
+
+/**
+ * @brief xi_time_event_move_to_the_end
+ *
+ * Helper function that moves the element form the given position ( index ) to the end of
+ * the vector so it is the last element in the vector.
+ *
+ * @param vector
+ * @param index
+ */
+void xi_time_event_move_to_the_end( xi_vector_t* vector, xi_vector_index_type_t index )
+{
+    assert( vector->elem_no > 0 );
+    assert( index < vector->elem_no );
+    assert( index >= 0 );
+
+    const xi_vector_index_type_t last_elem_index = vector->elem_no - 1;
+
+    xi_vector_index_type_t elem_to_swap_with_index = index + 1;
+
+    while ( index < last_elem_index )
+    {
+        xi_swap_time_events( vector, index, elem_to_swap_with_index );
+
+        /* once swapped update the indexes */
+        index += 1;
+        elem_to_swap_with_index += 1;
+    }
+}
+
+/**
+ * @brief xi_insert_time_event
+ *
+ * Helper function that inserts the new time event element to the vector.
+ *
+ * @param vector
+ * @param time_event
+ */
+static const xi_vector_elem_t*
+xi_insert_time_event( xi_vector_t* vector, xi_time_event_t* time_event )
+{
+    xi_state_t local_state = XI_STATE_OK;
+
+    /* add the element to the end of the vector */
+    {
+        const xi_vector_elem_t* inserted_element = xi_vector_push(
+            vector, XI_VEC_CONST_VALUE_PARAM( XI_VEC_VALUE_PTR( time_event ) ) );
+
+        XI_CHECK_MEMORY( inserted_element, local_state );
+    }
+
+    /* update the time event new position */
+    time_event->position = vector->elem_no - 1;
+
+    xi_vector_index_type_t index =
+        xi_time_event_bubble_and_sort_down( vector, vector->elem_no - 1 );
+
+    const xi_vector_elem_t* element_added = &vector->array[index];
+
+    return element_added;
+
+err_handling:
+    return NULL;
+}
+
+/**
+ * @brief xi_time_event_clean_time_event
+ *
+ * Helper function that cleans the time event so that it is unusable
+ *
+ * @param time_event
+ */
+static void xi_time_event_dispose_time_event( xi_time_event_t* time_event )
+{
+    if ( NULL != time_event->time_event_handle )
+    {
+        time_event->time_event_handle->position = NULL;
+    }
+}
+
+xi_state_t xi_time_event_add( xi_vector_t* vector,
```

Comment:
Perhaps a comment block like 

/*
 * STATIC INTERNAL FUNCTIONS
 */

and 

/* 
 * Public Functions
 */

to make some of thes ecomments and lack of comments more obvious.